File: sdks/dotnet/src/Linktor/Utils/WebhookVerifier.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Linktor.Types;

namespace Linktor.Utils;

public static class WebhookVerifier
{
    public static string ComputeSignature(byte[] payload, string secret)
    {
        using var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(secret));
        var hash = hmac.ComputeHash(payload);
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    public static string ComputeSignature(string payload, string secret)
        => ComputeSignature(Encoding.UTF8.GetBytes(payload), secret);

    public static bool VerifySignature(byte[] payload, string signature, string secret)
    {
        if (string.IsNullOrEmpty(signature) || string.IsNullOrEmpty(secret))
            return false;

        var expected = ComputeSignature(payload, secret);
        return CryptographicOperations.FixedTimeEquals(
            Encoding.UTF8.GetBytes(signature.ToLowerInvariant()),
            Encoding.UTF8.GetBytes(expected)
        );
    }

    public static bool VerifySignature(string payload, string signature, string secret)
        => VerifySignature(Encoding.UTF8.GetBytes(payload), signature, secret);

    public static bool Verify(byte[] payload, Dictionary<string, string> headers, string secret, int? toleranceSeconds = null)
    {
        var tolerance = toleranceSeconds ?? WebhookConstants.DefaultToleranceSeconds;

        // Get signature (case-insensitive)
        var signature = headers.TryGetValue(WebhookConstants.SignatureHeader, out var sig) ? sig
            : headers.TryGetValue(WebhookConstants.SignatureHeader.ToLower(), out sig) ? sig
            : null;

        if (string.IsNullOrEmpty(signature))
            return false;

        // Verify timestamp if present
        var timestampStr = headers.TryGetValue(WebhookConstants.TimestampHeader, out var ts) ? ts
            : headers.TryGetValue(WebhookConstants.TimestampHeader.ToLower(), out ts) ? ts
            : null;

        if (!string.IsNullOrEmpty(timestampStr))
        {
            if (!long.TryParse(timestampStr, out var timestamp))
                return false;

            var now = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
            if (Math.Abs(now - timestamp) > tolerance)
                return false;
        }

        return VerifySignature(payload, signature, secret);
    }

    public static WebhookEvent ConstructEvent(byte[] payload, Dictionary<string, string> headers, string secret, int? toleranceSeconds = null)
    {
        var tolerance = toleranceSeconds == 0 ? WebhookConstants.DefaultToleranceSeconds
            : toleranceSeconds ?? WebhookConstants.DefaultToleranceSeconds;

        if (!Verify(payload, headers, secret, tolerance))
            throw new WebhookVerificationException("Webhook signature verification failed");

        try
        {
            var json = Encoding.UTF8.GetString(payload);
            var ev = JsonSerializer.Deserialize<WebhookEvent>(json, new JsonSerializerOptions
            {
                PropertyNamingPolicy = JsonNamingPolicy.CamelCase
            });

            if (ev == null || string.IsNullOrEmpty(ev.Id) || string.IsNullOrEmpty(ev.Type))
                throw new WebhookVerificationException("Invalid webhook event structure");

            return ev;
        }
        catch (JsonException ex)
        {
            throw new WebhookVerificationException($"Failed to parse webhook event: {ex.Message}");
        }
    }

    public static WebhookEvent ConstructEvent(string payload, Dictionary<string, string> headers, string secret, int? toleranceSeconds = null)
        => ConstructEvent(Encoding.UTF8.GetBytes(payload), headers, secret, toleranceSeconds);
}

File: sdks/java/src/main/java/io/linktor/utils/WebhookVerifier.java
package io.linktor.utils;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializer;
import io.linktor.types.Webhook;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Instant;
import java.util.Map;

public class WebhookVerifier {
    private static final String HMAC_SHA256 = "HmacSHA256";
    private static final Gson gson = new GsonBuilder()
            .registerTypeAdapter(Instant.class, (JsonDeserializer<Instant>) (json, typeOfT, context) ->
                    Instant.parse(json.getAsString()))
            .create();

    /**
     * Compute HMAC-SHA256 signature for the given payload
     */
    public static String computeSignature(byte[] payload, String secret) {
        try {
            Mac mac = Mac.getInstance(HMAC_SHA256);
            SecretKeySpec secretKey = new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), HMAC_SHA256);
            mac.init(secretKey);
            byte[] hash = mac.doFinal(payload);
            return bytesToHex(hash);
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            throw new RuntimeException("Failed to compute signature", e);
        }
    }

    /**
     * Compute HMAC-SHA256 signature for the given payload string
     */
    public static String computeSignature(String payload, String secret) {
        return computeSignature(payload.getBytes(StandardCharsets.UTF_8), secret);
    }

    /**
     * Verify webhook signature only (no timestamp validation)
     */
    public static boolean verifySignature(byte[] payload, String signature, String secret) {
        if (signature == null || signature.isEmpty() || secret == null || secret.isEmpty()) {
            return false;
        }

        String expected = computeSignature(payload, secret);
        return MessageDigest.isEqual(
                signature.getBytes(StandardCharsets.UTF_8),
                expected.getBytes(StandardCharsets.UTF_8)
        );
    }

    /**
     * Verify webhook signature only (no timestamp validation)
     */
    public static boolean verifySignature(String payload, String signature, String secret) {
        return verifySignature(payload.getBytes(StandardCharsets.UTF_8), signature, secret);
    }

    /**
     * Verify webhook with signature and timestamp validation
     */
    public static boolean verify(byte[] payload, Map<String, String> headers, String secret, int toleranceSeconds) {
        String signature = headers.get(Webhook.SIGNATURE_HEADER);
        if (signature == null) {
            signature = headers.get(Webhook.SIGNATURE_HEADER.toLowerCase());
        }

        if (signature == null || signature.isEmpty()) {
            return false;
        }

        // Verify timestamp if present
        String timestampStr = headers.get(Webhook.TIMESTAMP_HEADER);
        if (timestampStr == null) {
            timestampStr = headers.get(Webhook.TIMESTAMP_HEADER.toLowerCase());
        }

        if (timestampStr != null && !timestampStr.isEmpty()) {
            try {
                long timestamp = Long.parseLong(timestampStr);
                long now = Instant.now().getEpochSecond();
                if (Math.abs(now - timestamp) > toleranceSeconds) {
                    return false;
                }
            } catch (NumberFormatException e) {
                return false;
            }
        }

        return verifySignature(payload, signature, secret);
    }

    /**
     * Verify webhook with signature and default timestamp tolerance (5 minutes)
     */
    public static boolean verify(byte[] payload, Map<String, String> headers, String secret) {
        return verify(payload, headers, secret, Webhook.DEFAULT_TOLERANCE_SECONDS);
    }

    /**
     * Construct and verify a webhook event
     */
    public static Webhook.WebhookEvent constructEvent(byte[] payload, Map<String, String> headers, String secret, int toleranceSeconds)
            throws LinktorException.WebhookVerificationException {
        if (!verify(payload, headers, secret, toleranceSeconds == 0 ? Webhook.DEFAULT_TOLERANCE_SECONDS : toleranceSeconds)) {
            throw new LinktorException.WebhookVerificationException("Webhook signature verification failed");
        }

        try {
            String json = new String(payload, StandardCharsets.UTF_8);
            Webhook.WebhookEvent event = gson.fromJson(json, Webhook.WebhookEvent.class);

            if (event.getId() == null || event.getType() == null) {
                throw new LinktorException.WebhookVerificationException("Invalid webhook event structure");
            }

            return event;
        } catch (Exception e) {
            if (e instanceof LinktorException.WebhookVerificationException) {
                throw (LinktorException.WebhookVerificationException) e;
            }
            throw new LinktorException.WebhookVerificationException("Failed to parse webhook event: " + e.getMessage());
        }
    }

    /**
     * Construct and verify a webhook event with default tolerance
     */
    public static Webhook.WebhookEvent constructEvent(byte[] payload, Map<String, String> headers, String secret)
            throws LinktorException.WebhookVerificationException {
        return constructEvent(payload, headers, secret, Webhook.DEFAULT_TOLERANCE_SECONDS);
    }

    /**
     * Construct and verify a webhook event from string payload
     */
    public static Webhook.WebhookEvent constructEvent(String payload, Map<String, String> headers, String secret)
            throws LinktorException.WebhookVerificationException {
        return constructEvent(payload.getBytes(StandardCharsets.UTF_8), headers, secret, Webhook.DEFAULT_TOLERANCE_SECONDS);
    }

    private static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
}
